### PR TITLE
Setup for case 3, flow benchmark 3d

### PR DESCRIPTION
**Pull request by jhabriel**
## Proposed changes

This PR introduces the setup for case 3, flow benchmark 3d, as a porepy example. Tests for the sanity of benchmark parameters specification (e.g., effective permeabilities and boundary conditions) are included. 

Idea for a proper test: Ideally, we should try to sample the reference value of the plot over line from the benchmark and compare them with the approximate pressure solutions. Might be a nice little side project for the user group.

## Types of changes

What types of changes does this PR introduce to PorePy?
_Put an `x` in the boxes that apply._

- [ ] Minor change (e.g., dependency bumps, broken links).
- [ ] Bugfix (non-breaking change which fixes an issue).
- [x] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).
- [x] Testing (contribution related to testing of existing or new functionality).
- [x] Documentation (contribution related to adding, improving, or fixing documentation).
- [ ] Maintenance (e.g., improve logic and performance, remove obsolete code).
- [ ] Other:

## Checklist

_Put an `x` in the boxes that apply or explain briefly why the box is not relevant._

- [x] The documentation is up-to-date.
- [x] Static typing is included in the update.
- [x] This PR does not duplicate existing functionality.
- [x] The update is covered by the test suite (including tests added in the PR).
- [x] If new skipped tests have been introduced in this PR, `pytest` was run with the `--run-skipped` flag.

### CODE REVIEW

File: tests/functional/test_benchmark_3d_case_3.py [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+    """
+    Test if the permeability values are consistent with the benchmark specification.
+
+    Parameters:
+        model: ModelWithEffectivePermeability
+            Solved model. Returned by the `model()` fixture.
+
+    """
+    for sd, d in model.mdg.subdomains(return_data=True):
+        val = np.mean(
```
Comment (IvarStefansson) [inline tests/functional/test_benchmark_3d_case_3.py:94]: Why do you use mean?

File: tests/functional/test_benchmark_3d_case_3.py [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+            d[pp.PARAMETERS][model.darcy_keyword]["second_order_tensor"].values[0][0]
+        )
+        if sd.dim == 3:
+            np.testing.assert_allclose(val, 1)
```
Comment (IvarStefansson) [inline tests/functional/test_benchmark_3d_case_3.py:98]: Please either explain the origin of these hard-coded values or (even better) define a method effective_tangential_permeability which computes them (using Eq. 6a, right?) from the values provided in table 5.

File: tests/functional/test_benchmark_3d_case_3.py [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+        the domain. At the outlet boundary, we check that the pressure value is zero.
+
+    """
+    bg, data_bg = model.mdg.boundaries(return_data=True, dim=2)[0]
+
+    # Inlet boundary
+    south_side = model.domain_boundary_sides(bg).south
+    inlet_flux = np.sum(data_bg["iterate_solutions"]["darcy_flux"][0][south_side])
+    np.testing.assert_allclose(inlet_flux, desired=-1/3, atol=1e-5)
+
```
Comment (IvarStefansson) [inline tests/functional/test_benchmark_3d_case_3.py:141]: Could you please comment (on GH) on the use of .testing? And do we really need allclose? Shouldn't isclose suffice?

File: tests/functional/test_benchmark_3d_case_3.py
Comment (IvarStefansson) [inline tests/functional/test_benchmark_3d_case_3.py:1]: Allthough I agree with you that a stricter test would be even better, I like the implementation a lot.

**Review by IvarStefansson — CHANGES_REQUESTED** [review]
Thanks! Nice work. I was pleasantly surprised how few LOC this ammounts to. A few relatively minor comments.

**Review by IvarStefansson — COMMENTED** [review]
Thanks! Nice work. I was pleasantly surprised how few LOC this ammounts to. A few relatively minor comments.

**Review by IvarStefansson — COMMENTED** [review]
Thanks! Nice work. I was pleasantly surprised how few LOC this ammounts to. A few relatively minor comments.

File: examples/flow_benchmark_3d_case_3.py [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+
+import porepy as pp
+from porepy.applications.md_grids.mdg_library import benchmark_3d_case_3
+from porepy.fracs.fracture_network_3d import FractureNetwork3d
+from porepy.models.constitutive_laws import DimensionDependentPermeability
+
+solid_constants = pp.SolidConstants(
+    {
+        "residual_aperture": 1e-2,
+        "normal_permeability": 1e4,
```
[the remark this answers is not in the document]
Comment (jhabriel) [reply examples/flow_benchmark_3d_case_3.py:30]: This is true. @keileg: You might benefit from checking these calculations:

https://www.dropbox.com/scl/fi/m7m7jxw9q3d2yboxb6z8y/permeability_calculations.pdf?rlkey=riq9vont0u3cznye0cgjsfqoi&dl=0

File: tests/functional/test_benchmark_3d_case_3.py (hunk 1 above)
Comment (jhabriel) [reply tests/functional/test_benchmark_3d_case_3.py:90]: Good point, `np.mean` is actually not needed here.

File: tests/functional/test_benchmark_3d_case_3.py (hunk 2 above)
Comment (jhabriel) [reply tests/functional/test_benchmark_3d_case_3.py:94]: I did not do that because the second order tensor stored in the data dictionary is actually the effective tangential permeability (the specific volume multiplies the tangential permeability [here](https://github.com/pmgbergen/porepy/blob/e4d484f915cde39cc17639e24c1ffe5ab37c1f99/src/porepy/models/fluid_mass_balance.py#L837)).

But this is not a problem, I will add a method `effective_tangential_permeability` to the `ModelWithEffectivePermeability` class with a more in-depth explanation.

File: tests/functional/test_benchmark_3d_case_3.py (hunk 3 above)
Comment (jhabriel) [reply tests/functional/test_benchmark_3d_case_3.py:135]: Yes, this is a scalar, `np.isclose()` suffices.

**Comment by jhabriel** [comment]
Thanks Ivar. I will work on incorporating the suggestions now.

File: tests/functional/test_benchmark_3d_case_3.py
Comment (jhabriel) [reply tests/functional/test_benchmark_3d_case_3.py:1]: Thanks. I wish we could have some tests on the produced results, but without comparing them to the reference solution, I don't really see how we can achieve this. I'm open to suggestions though.

File: tests/functional/test_benchmark_3d_case_3.py (hunk 2 above)
Comment (jhabriel) [reply tests/functional/test_benchmark_3d_case_3.py]: I have added the method `effective_tangential_permeability` as explained above. Please, let me know if this is more or less what you envisioned.

**Comment by jhabriel** [comment]
@IvarStefansson: I believe I've incorporated all your suggestions now.

File: tests/functional/test_benchmark_3d_case_3.py
Comment (IvarStefansson) [reply tests/functional/test_benchmark_3d_case_3.py:1]: I'm content as is. @keileg, please make a decision.

File: tests/functional/test_benchmark_3d_case_3.py [hunk 5]
```diff
@@ -81,59 +117,61 @@ def model() -> ModelWithEffectivePermeability:
 def test_effective_tangential_permeability_values(model) -> None:
     """Test if the permeability values are consistent with the benchmark specification.
 
+    The values are specified in Table 5 from [1]. We expect a value of normal
+    permeability = 2E6 for 2d interfaces, and 2e4 for 1d interfaces.
```
Comment (IvarStefansson) [inline tests/functional/test_benchmark_3d_case_3.py:121]: This comment seems to be misplaced.

**Review by IvarStefansson — COMMENTED** [review]
Looks good!

File: tests/functional/test_benchmark_3d_case_3.py (hunk 5 above)
Comment (jhabriel) [reply tests/functional/test_benchmark_3d_case_3.py]: Fixed.

**Review by IvarStefansson — COMMENTED** [review]
I'm happy. Over to you for final review, @keileg.

File: examples/flow_benchmark_3d_case_3.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (keileg) [reply examples/flow_benchmark_3d_case_3.py:30]: I agree, the values make sense.

File: tests/functional/test_benchmark_3d_case_3.py
Comment (keileg) [reply tests/functional/test_benchmark_3d_case_3.py:1]: I think this is a fair compromise.

**Review by keileg — COMMENTED** [review]
Thanks a lot for the contribution, @jhabriel. The files are fine with me, but let us come back to the placement (as you know, where to put such functionality has been a long standing issue). Just to be sure, I don't formally approve the PR yet, but there should be no need for any action from your side.

**Review by IvarStefansson — APPROVED** [review]
